pkg/querier/worker/worker_test.go

```diff
@@ -168,6 +197,67 @@ func TestResetConcurrency(t *testing.T) {
 	}
 }
 
+func TestQuerierWorker_getDesiredConcurrency(t *testing.T) {
```

Comment:
I would like to see a test case where `maxConcurrent < numInUse`. 
In that case `concurrency := w.cfg.MaxConcurrentRequests / numInUse` will be `0` and there's no guarrantee that some querier will connect to every scheduler.

In particular, I'd just try to ensure that every instance gets at least 1 connection.

Was this also the previous behavior?

Reply:
I think it's already covered by "should create 1 connection for each instance if max concurrency is set to 0", but I've added an additional explicit one in https://github.com/grafana/mimir/pull/3005/commits/91071134f22245506d543e8f63b789c4ed2dca55.

> In particular, I'd just try to ensure that every instance gets at least 1 connection.
> Was this also the previous behavior?

Yes.

Reply:
>> In particular, I'd just try to ensure that every instance gets at least 1 connection.
Was this also the previous behavior?
> Yes.

See: https://github.com/grafana/mimir/blob/12db6c06e6eb47ef89dcdb43acd4d23c516b8d84/pkg/querier/worker/worker.go#L264-L270

Reply:
The comment "This is dangerous b/c we may start exceeding PromQL max concurrency." has been removed because it's not true anymore (we don't limit PromQL engine concurrency anymore).

Reply:
> I think it's already covered by "should create 1 connection for each instance if max concurrency is set to 0"

Oh, right, sorry. I thought that was somehow a special case. Thank you for adding the testcase anyway.

---

Given there's already one connection ensured to each instance, we don't really need to ensure it at two levels (in `getDesiredConcurrency` and in `resetConcurrency`). But not sure if worth the change.